File: codegenTesting/sdr/dom/components/props/java/src/props/props.java
/*
 * This file is protected by Copyright. Please refer to the COPYRIGHT file
 * distributed with this source distribution.
 *
 * This file is part of REDHAWK codegenTesting.
 *
 * REDHAWK codegenTesting is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * REDHAWK codegenTesting is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 */
package props;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Properties;
import org.omg.CORBA.ORB;
import org.omg.PortableServer.POA;
import org.omg.PortableServer.POAPackage.ServantNotActive;
import org.omg.PortableServer.POAPackage.WrongPolicy;
import org.omg.CosNaming.NamingContextPackage.CannotProceed;
import org.omg.CosNaming.NamingContextPackage.InvalidName;
import org.omg.CosNaming.NamingContextPackage.NotFound;
import CF.PropertiesHolder;
import CF.ResourceHelper;
import CF.UnknownProperties;
import CF.LifeCyclePackage.InitializeError;
import CF.LifeCyclePackage.ReleaseError;
import CF.InvalidObjectReference;
import CF.PropertySetPackage.InvalidConfiguration;
import CF.PropertySetPackage.PartialConfiguration;
import CF.ResourcePackage.StartError;
import CF.ResourcePackage.StopError;
import CF.DataType;
import org.omg.CORBA.UserException;
import org.omg.CosNaming.NameComponent;
import org.apache.log4j.Logger;
import org.ossie.component.*;
import org.ossie.properties.*;

/**
 * This is the component code. This file contains all the access points
 * you need to use to be able to access all input and output ports,
 * respond to incoming data, and perform general component housekeeping
 *
 * Source: props.spd.xml
 *
 * @generated
 */
public class props extends props_base {
    public props()
    {
        super();
        this.stringSimple.addChangeListener(new PropertyListener<String>(){
                                       public void valueChanged(String old_value, String new_value){
                                           stringSimpleChanged(old_value, new_value);
                                       }
        });
        this.boolSimple.addChangeListener(new PropertyListener<Boolean>(){
                                       public void valueChanged(Boolean old_value, Boolean new_value){
                                           boolSimpleChanged(old_value, new_value);
                                       }
        });
        this.ulongSimple.addChangeListener(new PropertyListener<Integer>(){
                                       public void valueChanged(Integer old_value, Integer new_value){
                                           ulongSimpleChanged(old_value, new_value);
                                       }
        });
        this.shortSimple.addChangeListener(new PropertyListener<Short>(){
                                       public void valueChanged(Short old_value, Short new_value){
                                           shortSimpleChanged(old_value, new_value);
                                       }
        });
        this.floatSimple.addChangeListener(new PropertyListener<Float>(){
                                       public void valueChanged(Float old_value, Float new_value){
                                           floatSimpleChanged(old_value, new_value);
                                       }
        });
        this.octetSimple.addChangeListener(new PropertyListener<Byte>(){
                                       public void valueChanged(Byte old_value, Byte new_value){
                                           octetSimpleChanged(old_value, new_value);
                                       }
        });
        this.charSimple.addChangeListener(new PropertyListener<Character>(){
                                       public void valueChanged(Character old_value, Character new_value){
                                           charSimpleChanged(old_value, new_value);
                                       }
        });
        this.ushortSimple.addChangeListener(new PropertyListener<Short>(){
                                       public void valueChanged(Short old_value, Short new_value){
                                           ushortSimpleChanged(old_value, new_value);
                                       }
        });
        this.doubleSimple.addChangeListener(new PropertyListener<Double>(){
                                       public void valueChanged(Double old_value, Double new_value){
                                           doubleSimpleChanged(old_value, new_value);
                                       }
        });
        this.longSimple.addChangeListener(new PropertyListener<Integer>(){
                                       public void valueChanged(Integer old_value, Integer new_value){
                                           longSimpleChanged(old_value, new_value);
                                       }
        });
        this.longlongSimple.addChangeListener(new PropertyListener<Long>(){
                                       public void valueChanged(Long old_value, Long new_value){
                                           longlongSimpleChanged(old_value, new_value);
                                       }
        });
        this.ulonglongSimple.addChangeListener(new PropertyListener<Long>(){
                                       public void valueChanged(Long old_value, Long new_value){
                                           ulonglongSimpleChanged(old_value, new_value);
                                       }
        });
        this.stringSeq.addChangeListener(new PropertyListener<List<String>>(){
                                       public void valueChanged(List<String> old_value, List<String> new_value){
                                           stringSeqChanged(old_value, new_value);
                                       }
        });
        this.boolSeq.addChangeListener(new PropertyListener<List<Boolean>>(){
                                       public void valueChanged(List<Boolean> old_value, List<Boolean> new_value){
                                           boolSeqChanged(old_value, new_value);
                                       }
        });
        this.ulongSeq.addChangeListener(new PropertyListener<List<Integer>>(){
                                       public void valueChanged(List<Integer> old_value, List<Integer> new_value){
                                           ulongSeqChanged(old_value, new_value);
                                       }
        });
        this.longSeq.addChangeListener(new PropertyListener<List<Integer>>(){
                                       public void valueChanged(List<Integer> old_value, List<Integer> new_value){
                                           longSeqChanged(old_value, new_value);
                                       }
        });
        this.shortSeq.addChangeListener(new PropertyListener<List<Short>>(){
                                       public void valueChanged(List<Short> old_value, List<Short> new_value){
                                           shortSeqChanged(old_value, new_value);
                                       }
        });
        this.floatSeq.addChangeListener(new PropertyListener<List<Float>>(){
                                       public void valueChanged(List<Float> old_value, List<Float> new_value){
                                           floatSeqChanged(old_value, new_value);
                                       }
        });
        this.octetSeq.addChangeListener(new PropertyListener<List<Byte>>(){
                                       public void valueChanged(List<Byte> old_value, List<Byte> new_value){
                                           octetSeqChanged(old_value, new_value);
                                       }
        });
        this.charSeq.addChangeListener(new PropertyListener<List<Character>>(){
                                       public void valueChanged(List<Character> old_value, List<Character> new_value){
                                           charSeqChanged(old_value, new_value);
                                       }
        });
        this.ushortSeq.addChangeListener(new PropertyListener<List<Short>>(){
                                       public void valueChanged(List<Short> old_value, List<Short> new_value){
                                           ushortSeqChanged(old_value, new_value);
                                       }
        });
        this.ulonglongSeq.addChangeListener(new PropertyListener<List<Long>>(){
                                       public void valueChanged(List<Long> old_value, List<Long> new_value){
                                           ulonglongSeqChanged(old_value, new_value);
                                       }
        });
        this.longlongSeq.addChangeListener(new PropertyListener<List<Long>>(){
                                       public void valueChanged(List<Long> old_value, List<Long> new_value){
                                           longlongSeqChanged(old_value, new_value);
                                       }
        });
        this.doubleSeq.addChangeListener(new PropertyListener<List<Double>>(){
                                       public void valueChanged(List<Double> old_value, List<Double> new_value){
                                           doubleSeqChanged(old_value, new_value);
                                       }
        });
        this.structProp.addChangeListener(new PropertyListener<structProp_struct>(){
                                       public void valueChanged(structProp_struct old_value, structProp_struct new_value){
                                           structPropChanged(old_value, new_value);
                                       }
        });
        this.structSeqProp.addChangeListener(new PropertyListener<List<structSeqStruct_struct>>(){
                                       public void valueChanged(List<structSeqStruct_struct> old_value, List<structSeqStruct_struct> new_value){
                                           structSeqPropChanged(old_value, new_value);
                                       }
        });
    }

    protected int serviceFunction() 
    {
        //begin-user-code
        //end-user-code
        
            //begin-user-code
            // Process data here
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
            }
            return NOOP;    
        //begin-user-code
        //end-user-code
    }

    private void stringSimpleChanged(String old_value, String new_value)
    {
        stringSimple.setValue(new_value.concat(new_value));
    }

    private void boolSimpleChanged(Boolean old_value, Boolean new_value)
    {
        if (new_value != null)
            boolSimple.setValue(!new_value);
    }

    private void ulongSimpleChanged(Integer old_value, Integer new_value)
    {
        ulongSimple.setValue(new_value * 2);
    }

    private void shortSimpleChanged(Short old_value, Short new_value)
    {
        shortSimple.setValue((short) (new_value * 2));
    }

    private void floatSimpleChanged(Float old_value, Float new_value)
    {
        floatSimple.setValue(new_value * 2);
    }

    private void octetSimpleChanged(Byte old_value, Byte new_value)
    {
        octetSimple.setValue((byte) (new_value * 2));
    }

    private void charSimpleChanged(Character old_value, Character new_value)
    {
        charSimple.setValue(Character.toUpperCase(new_value));
    }

    private void ushortSimpleChanged(Short old_value, Short new_value)
    {
        ushortSimple.setValue(new_value * 2);
    }

    private void doubleSimpleChanged(Double old_value, Double new_value)
    {
        doubleSimple.setValue(new_value * 2);
    }

    private void longSimpleChanged(Integer old_value, Integer new_value)
    {
        longSimple.setValue(new_value * 2);
    }

    private void longlongSimpleChanged(Long old_value, Long new_value)
    {
        longlongSimple.setValue(new_value * 2);
    }

    private void ulonglongSimpleChanged(Long old_value, Long new_value)
    {
        ulonglongSimple.setValue(Integer.parseInt(new_value.toString()) * 2);
    }

    private void stringSeqChanged(List<String> old_value, List<String> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void boolSeqChanged(List<Boolean> old_value, List<Boolean> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void ulongSeqChanged(List<Integer> old_value, List<Integer> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void shortSeqChanged(List<Short> old_value, List<Short> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void floatSeqChanged(List<Float> old_value, List<Float> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void octetSeqChanged(List<Byte> old_value, List<Byte> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void charSeqChanged(List<Character> old_value, List<Character> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void ushortSeqChanged(List<Short> old_value, List<Short> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void ulonglongSeqChanged(List<Long> old_value, List<Long> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void longSeqChanged(List<Integer> old_value, List<Integer> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void longlongSeqChanged(List<Long> old_value, List<Long> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void doubleSeqChanged(List<Double> old_value, List<Double> new_value)
    {
        java.util.Collections.reverse(new_value);
    }

    private void structPropChanged(structProp_struct old_value, structProp_struct new_value)
    {
        try {
            structProp_struct value = new_value;
            value.structStringSimple.setValue(value.structStringSimple.getValue().concat(value.structStringSimple.getValue()));
            value.structBoolSimple.setValue(!value.structBoolSimple.getValue());
            value.structUlongSimple.setValue(value.structUlongSimple.getValue() * 2);
            value.structShortSimple.setValue((short) (value.structShortSimple.getValue() * 2));
            value.structFloatSimple.setValue(value.structFloatSimple.getValue() * 2);
            value.structOctetSimple.setValue((byte) (value.structOctetSimple.getValue() * 2));
            value.structCharSimple.setValue(Character.toUpperCase(value.structCharSimple.getValue()));
            value.structUshortSimple.setValue(value.structUshortSimple.getValue() * 2);
            value.structDoubleSimple.setValue(value.structDoubleSimple.getValue() * 2);
            value.structLongSimple.setValue(value.structLongSimple.getValue() * 2);
            value.structLonglongSimple.setValue(value.structLonglongSimple.getValue() * 2);
            value.structUlonglongSimple.setValue(Integer.parseInt(value.structUlonglongSimple.getValue().toString()) * 2);
        } catch(Exception e) {
            //pass
        }
    }

    private void structSeqPropChanged(List<structSeqStruct_struct> old_value, List<structSeqStruct_struct> new_value)
    {
        for (structSeqStruct_struct value : new_value) {
            value.structSeqBoolSimple.setValue(!value.structSeqBoolSimple.getValue());
            value.structSeqFloatSimple.setValue(value.structSeqFloatSimple.getValue() * 2);
            value.structSeqStringSimple.setValue(value.structSeqStringSimple.getValue().concat(value.structSeqStringSimple.getValue()));
            value.structSeqShortSimple.setValue((short) (value.structSeqShortSimple.getValue() * 2));
        }
    }

    /**
     * Set additional options for ORB startup. For example:
     *
     *   orbProps.put("com.sun.CORBA.giop.ORBFragmentSize", Integer.toString(fragSize));
     *
     * @param orbProps
     */
    public static void configureOrb(final Properties orbProps) {
    }
}

File: codegenTesting/sdr/dom/components/props/cpp/props.cpp
/*
 * This file is protected by Copyright. Please refer to the COPYRIGHT file
 * distributed with this source distribution.
 *
 * This file is part of REDHAWK codegenTesting.
 *
 * REDHAWK codegenTesting is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * REDHAWK codegenTesting is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 */
/**************************************************************************

    This is the component code. This file contains the child class where
    custom functionality can be added to the component. Custom
    functionality to the base class can be extended here. Access to
    the ports can also be done from this class

**************************************************************************/

#include "props.h"

PREPARE_LOGGING(props_i)

props_i::props_i(const char *uuid, const char *label) :
    props_base(uuid, label)
{
    old_bool = false;
    setPropertyChangeListener("stringSimple", this, &props_i::stringSimpleChanged);
    setPropertyChangeListener("boolSimple", this, &props_i::boolSimpleChanged);
    setPropertyChangeListener("ulongSimple", this, &props_i::ulongSimpleChanged);
    setPropertyChangeListener("shortSimple", this, &props_i::shortSimpleChanged);
    setPropertyChangeListener("floatSimple", this, &props_i::floatSimpleChanged);
    setPropertyChangeListener("octetSimple", this, &props_i::octetSimpleChanged);
    setPropertyChangeListener("charSimple", this, &props_i::charSimpleChanged);
    setPropertyChangeListener("ushortSimple", this, &props_i::ushortSimpleChanged);
    setPropertyChangeListener("doubleSimple", this, &props_i::doubleSimpleChanged);
    setPropertyChangeListener("longSimple", this, &props_i::longSimpleChanged);
    setPropertyChangeListener("longlongSimple", this, &props_i::longlongSimpleChanged);
    setPropertyChangeListener("ulonglongSimple", this, &props_i::ulonglongSimpleChanged);
    setPropertyChangeListener("stringSeq", this, &props_i::stringSeqChanged);
    setPropertyChangeListener("boolSeq", this, &props_i::boolSeqChanged);
    setPropertyChangeListener("ulongSeq", this, &props_i::ulongSeqChanged);
    setPropertyChangeListener("shortSeq", this, &props_i::shortSeqChanged);
    setPropertyChangeListener("floatSeq", this, &props_i::floatSeqChanged);
    setPropertyChangeListener("octetSeq", this, &props_i::octetSeqChanged);
    setPropertyChangeListener("charSeq", this, &props_i::charSeqChanged);
    setPropertyChangeListener("ushortSeq", this, &props_i::ushortSeqChanged);
    setPropertyChangeListener("doubleSeq", this, &props_i::doubleSeqChanged);
    setPropertyChangeListener("longSeq", this, &props_i::longSeqChanged);
    setPropertyChangeListener("longlongSeq", this, &props_i::longlongSeqChanged);
    setPropertyChangeListener("ulonglongSeq", this, &props_i::ulonglongSeqChanged);
    setPropertyChangeListener("structProp", this, &props_i::structPropChanged);
    setPropertyChangeListener("structSeqProp", this, &props_i::structSeqPropChanged);
}

props_i::~props_i()
{
}

void props_i::stringSimpleChanged (const std::string& id)
{
    stringSimple = stringSimple + stringSimple;
}

void props_i::boolSimpleChanged (const std::string& id)
{
    if (!old_bool) {
        boolSimple = !boolSimple;
        old_bool = true;
    } else {
        old_bool = false;
    }
}

void props_i::ulongSimpleChanged (const std::string& id)
{
    ulongSimple = ulongSimple * 2;
}

void props_i::shortSimpleChanged (const std::string& id)
{
    shortSimple = shortSimple * 2;
}

void props_i::floatSimpleChanged (const std::string& id)
{
    floatSimple = floatSimple * 2;
}

void props_i::octetSimpleChanged (const std::string& id)
{
    octetSimple = octetSimple * 2;
}

void props_i::charSimpleChanged (const std::string& id)
{
    charSimple = std::toupper(charSimple);
}

void props_i::ushortSimpleChanged (const std::string& id)
{
    ushortSimple = ushortSimple * 2;
}

void props_i::doubleSimpleChanged (const std::string& id)
{
    doubleSimple = doubleSimple * 2;
}

void props_i::longSimpleChanged (const std::string& id)
{
    longSimple = longSimple * 2;
}

void props_i::longlongSimpleChanged (const std::string& id)
{
    longlongSimple = longlongSimple * 2;
}

void props_i::ulonglongSimpleChanged (const std::string& id)
{
    ulonglongSimple = ulonglongSimple * 2;
}

void props_i::stringSeqChanged (const std::string& id)
{
    std::vector<std::string> _stringSeq;

    for(int i = stringSeq.size()-1; i >= 0; i--){
        _stringSeq.push_back(stringSeq.at(i));
    }

    stringSeq.clear();

    for(unsigned int i = 0; i < _stringSeq.size(); i++){
        stringSeq.push_back(_stringSeq.at(i));
    }
}

void props_i::boolSeqChanged (const std::string& id)
{
    std::vector<bool> _boolSeq;

    for(int i = boolSeq.size()-1; i >= 0; i--){
        _boolSeq.push_back(boolSeq.at(i));
    }

    boolSeq.clear();

    for(unsigned int i = 0; i < _boolSeq.size(); i++){
        boolSeq.push_back(_boolSeq.at(i));
    }
}

void props_i::ulongSeqChanged (const std::string& id)
{
    std::vector<CORBA::ULong> _ulongSeq;

    for(int i = ulongSeq.size()-1; i >= 0; i--){
        _ulongSeq.push_back(ulongSeq.at(i));
    }

    ulongSeq.clear();

    for(unsigned int i = 0; i < _ulongSeq.size(); i++){
        ulongSeq.push_back(_ulongSeq.at(i));
    }
}

void props_i::shortSeqChanged (const std::string& id)
{
    std::vector<CORBA::Short> _shortSeq;

    for(int i = shortSeq.size()-1; i >= 0; i--){
        _shortSeq.push_back(shortSeq.at(i));
    }

    shortSeq.clear();

    for(unsigned int i = 0; i < _shortSeq.size(); i++){
        shortSeq.push_back(_shortSeq.at(i));
    }
}

void props_i::floatSeqChanged (const std::string& id)
{
    std::vector<CORBA::Float> _floatSeq;

    for(int i = floatSeq.size()-1; i >= 0; i--){
        _floatSeq.push_back(floatSeq.at(i));
    }

    floatSeq.clear();

    for(unsigned int i = 0; i < _floatSeq.size(); i++){
        floatSeq.push_back(_floatSeq.at(i));
    }
}

void props_i::octetSeqChanged (const std::string& id)
{
    std::vector<CORBA::Octet> _octetSeq;

    for(int i = octetSeq.size()-1; i >= 0; i--){
        _octetSeq.push_back(octetSeq.at(i));
    }

    octetSeq.clear();

    for(unsigned int i = 0; i < _octetSeq.size(); i++){
        octetSeq.push_back(_octetSeq.at(i));
    }
}

void props_i::charSeqChanged (const std::string& id)
{
    std::vector<char> _charSeq;

    for(int i = charSeq.size()-1; i >= 0; i--){
        _charSeq.push_back(charSeq.at(i));
    }

    charSeq.clear();

    for(unsigned int i = 0; i < _charSeq.size(); i++){
        charSeq.push_back(_charSeq.at(i));
    }
}

void props_i::ushortSeqChanged (const std::string& id)
{
    std::vector<CORBA::UShort> _ushortSeq;

    for(int i = ushortSeq.size()-1; i >= 0; i--){
        _ushortSeq.push_back(ushortSeq.at(i));
    }

    ushortSeq.clear();

    for(unsigned int i = 0; i < _ushortSeq.size(); i++){
        ushortSeq.push_back(_ushortSeq.at(i));
    }
}

void props_i::doubleSeqChanged (const std::string& id)
{
    std::vector<CORBA::Double> _doubleSeq;

    for(int i = doubleSeq.size()-1; i >= 0; i--){
        _doubleSeq.push_back(doubleSeq.at(i));
    }

    doubleSeq.clear();

    for(unsigned int i = 0; i < _doubleSeq.size(); i++){
        doubleSeq.push_back(_doubleSeq.at(i));
    }
}

void props_i::longSeqChanged (const std::string& id)
{
    std::vector<CORBA::Long> _longSeq;

    for(int i = longSeq.size()-1; i >= 0; i--){
        _longSeq.push_back(longSeq.at(i));
    }

    longSeq.clear();

    for(unsigned int i = 0; i < _longSeq.size(); i++){
        longSeq.push_back(_longSeq.at(i));
    }
}

void props_i::longlongSeqChanged (const std::string& id)
{
    std::vector<CORBA::LongLong> _longlongSeq;

    for(int i = longlongSeq.size()-1; i >= 0; i--){
        _longlongSeq.push_back(longlongSeq.at(i));
    }

    longlongSeq.clear();

    for(unsigned int i = 0; i < _longlongSeq.size(); i++){
        longlongSeq.push_back(_longlongSeq.at(i));
    }
}

void props_i::ulonglongSeqChanged (const std::string& id)
{
    std::vector<CORBA::ULongLong> _ulonglongSeq;

    for(int i = ulonglongSeq.size()-1; i >= 0; i--){
        _ulonglongSeq.push_back(ulonglongSeq.at(i));
    }

    ulonglongSeq.clear();

    for(unsigned int i = 0; i < _ulonglongSeq.size(); i++){
        ulonglongSeq.push_back(_ulonglongSeq.at(i));
    }
}

void props_i::structPropChanged (const std::string& id)
{
    try{
        structProp.structBoolSimple = !structProp.structBoolSimple;
        structProp.structDoubleSimple = structProp.structDoubleSimple * 2;
        structProp.structFloatSimple = structProp.structFloatSimple * 2;
        structProp.structLongSimple = structProp.structLongSimple * 2;
        structProp.structLonglongSimple = structProp.structLonglongSimple * 2;
        structProp.structOctetSimple = structProp.structOctetSimple * 2;
        structProp.structShortSimple = structProp.structShortSimple * 2;
        structProp.structStringSimple = structProp.structStringSimple + structProp.structStringSimple;
        structProp.structUlongSimple = structProp.structUlongSimple * 2;
        structProp.structUlonglongSimple = structProp.structUlonglongSimple * 2;
        structProp.structUshortSimple = structProp.structUshortSimple * 2;
        structProp.structCharSimple = std::toupper(structProp.structCharSimple);
    } catch (std::exception e) {
        //pass
    }
}

void props_i::structSeqPropChanged (const std::string& id)
{
    for(unsigned int i = 0; i < structSeqProp.size(); i++){
        structSeqProp.at(i).structSeqBoolSimple = !structSeqProp.at(i).structSeqBoolSimple;
        structSeqProp.at(i).structSeqFloatSimple = structSeqProp.at(i).structSeqFloatSimple * 2;
        structSeqProp.at(i).structSeqShortSimple = structSeqProp.at(i).structSeqShortSimple * 2;
        structSeqProp.at(i).structSeqStringSimple = structSeqProp.at(i).structSeqStringSimple + structSeqProp.at(i).structSeqStringSimple;
    }
}

/***********************************************************************************************

    Basic functionality:

        The service function is called by the serviceThread object (of type ProcessThread).
        This call happens immediately after the previous call if the return value for
        the previous call was NORMAL.
        If the return value for the previous call was NOOP, then the serviceThread waits
        an amount of time defined in the serviceThread's constructor.
        
    SRI:
        To create a StreamSRI object, use the following code:
        	stream_id = "";
	    	sri = BULKIO::StreamSRI();
	    	sri.hversion = 1;
	    	sri.xstart = 0.0;
	    	sri.xdelta = 0.0;
	    	sri.xunits = BULKIO::UNITS_TIME;
	    	sri.subsize = 0;
	    	sri.ystart = 0.0;
	    	sri.ydelta = 0.0;
	    	sri.yunits = BULKIO::UNITS_NONE;
	    	sri.mode = 0;
	    	sri.streamID = this->stream_id.c_str();

	Time:
	    To create a PrecisionUTCTime object, use the following code:
	        struct timeval tmp_time;
	        struct timezone tmp_tz;
	        gettimeofday(&tmp_time, &tmp_tz);
	        double wsec = tmp_time.tv_sec;
	        double fsec = tmp_time.tv_usec / 1e6;;
	        BULKIO::PrecisionUTCTime tstamp = BULKIO::PrecisionUTCTime();
	        tstamp.tcmode = BULKIO::TCM_CPU;
	        tstamp.tcstatus = (short)1;
	        tstamp.toff = 0.0;
	        tstamp.twsec = wsec;
	        tstamp.tfsec = fsec;
        
    Ports:

        Data is passed to the serviceFunction through the getPacket call (BULKIO only).
        The dataTransfer class is a port-specific class, so each port implementing the
        BULKIO interface will have its own type-specific dataTransfer.

        The argument to the getPacket function is a floating point number that specifies
        the time to wait in seconds. A zero value is non-blocking. A negative value
        is blocking.

        Each received dataTransfer is owned by serviceFunction and *MUST* be
        explicitly deallocated.

        To send data using a BULKIO interface, a convenience interface has been added 
        that takes a std::vector as the data input

        NOTE: If you have a BULKIO dataSDDS port, you must manually call 
              "port->updateStats()" to update the port statistics when appropriate.

        Example:
            // this example assumes that the component has two ports:
            //  A provides (input) port of type BULKIO::dataShort called short_in
            //  A uses (output) port of type BULKIO::dataFloat called float_out
            // The mapping between the port and the class is found
            // in the component base class header file

            BULKIO_dataShort_In_i::dataTransfer *tmp = short_in->getPacket(-1);
            if (not tmp) { // No data is available
                return NOOP;
            }

            std::vector<float> outputData;
            outputData.resize(tmp->dataBuffer.size());
            for (unsigned int i=0; i<tmp->dataBuffer.size(); i++) {
                outputData[i] = (float)tmp->dataBuffer[i];
            }

            // NOTE: You must make at least one valid pushSRI call
            if (tmp->sriChanged) {
                float_out->pushSRI(tmp->SRI);
            }
            float_out->pushPacket(outputData, tmp->T, tmp->EOS, tmp->streamID);

            delete tmp; // IMPORTANT: MUST RELEASE THE RECEIVED DATA BLOCK
            return NORMAL;

        Interactions with non-BULKIO ports are left up to the component developer's discretion

    Properties:
        
        Properties are accessed directly as member variables. For example, if the
        property name is "baudRate", it may be accessed within member functions as
        "baudRate". Unnamed properties are given a generated name of the form
        "prop_n", where "n" is the ordinal number of the property in the PRF file.
        Property types are mapped to the nearest C++ type, (e.g. "string" becomes
        "std::string"). All generated properties are declared in the base class
        (props_base).
    
        Simple sequence properties are mapped to "std::vector" of the simple type.
        Struct properties, if used, are mapped to C++ structs defined in the
        generated file "struct_props.h". Field names are taken from the name in
        the properties file; if no name is given, a generated name of the form
        "field_n" is used, where "n" is the ordinal number of the field.
        
        Example:
            // This example makes use of the following Properties:
            //  - A float value called scaleValue
            //  - A boolean called scaleInput
              
            if (scaleInput) {
                dataOut[i] = dataIn[i] * scaleValue;
            } else {
                dataOut[i] = dataIn[i];
            }
            
        A callback method can be associated with a property so that the method is
        called each time the property value changes.  This is done by calling 
        setPropertyChangeListener(<property name>, this, &props_i::<callback method>)
        in the constructor.
            
        Example:
            // This example makes use of the following Properties:
            //  - A float value called scaleValue
            
        //Add to props.cpp
        props_i::props_i(const char *uuid, const char *label) :
            props_base(uuid, label)
        {
            setPropertyChangeListener("scaleValue", this, &props_i::scaleChanged);
        }

        void props_i::scaleChanged(const std::string& id){
            std::cout << "scaleChanged scaleValue " << scaleValue << std::endl;
        }
            
        //Add to props.h
        void scaleChanged(const std::string&);
        
        
************************************************************************************************/
int props_i::serviceFunction()
{
    LOG_DEBUG(props_i, "serviceFunction() example log message");
    
    return NOOP;
}
